_drafts/practical_zephyr_05_west.md

```diff
@@ -0,0 +1,1246 @@
+$ tree --dirsfirst -a -L 4 --filelimit 10 -I hal
+.
+├── .west
+│   └── config
+├── app
+├── deps
+│   ├── modules
+│   └── zephyr
+└── shared
+    └── dummy
+        └── test.txt
+```
+
+Without checking out a new branch, I can modify the contents of `test.txt` and run `west update`. The changes won't be lost and `west update` succeeds since the changes are not in conflict with any incoming change:
+
+```bash
+$ echo 123-test >> shared/dummy/test.txt
+$ west update
+...
+=== updating dummy (shared/dummy):
+--- dummy: fetching, need revision main
+From github.com:lmapii/practical-zephyr-t2-dummy
+ * branch            main       -> FETCH_HEAD
+M       test.txt
+HEAD is now at c8deb58 initial commit
+...
+
+$ cat shared/dummy/test.txt
+123-test
+```
+
+The output of `west update` shows me that I have in fact local changes to `test.txt`, but it doesn't reset the file. The changes are not lost.
+
+Since we're using `main` as `revision` for this dependency (e.g., since we're still very much in development), it _could_ happen that someone else is writing to `test.txt`, e.g., the value `not-a-test.` (let's not discuss using development branches etc., I'm just showing _West's_ behavior). Trying to run `west update` with an incoming change leads to an error - the local changes are still not lost:
+
+```bash
+$ west update
+...
+=== updating dummy (shared/dummy):
+--- dummy: fetching, need revision main
+remote: Enumerating objects: 5, done.
+remote: Counting objects: 100% (5/5), done.
+remote: Total 3 (delta 0), reused 3 (delta 0), pack-reused 0
+Unpacking objects: 100% (3/3), 229 bytes | 57.00 KiB/s, done.
+From github.com:lmapii/practical-zephyr-t2-dummy
+ * branch            main       -> FETCH_HEAD
+error: Your local changes to the following files would be overwritten by checkout:
+        test.txt
+Please commit your changes or stash them before you switch branches.
+Aborting
+...
+ERROR: update failed for project dummy
+```
+
+The only irreversible mistake that we can make is to commit changes to the `manifest-rev` branch. Let's do just that:
+
+```bash
+$ cd shared/dummy
+$ git add test.txt
+$ git commit -m "an honest mistake"
+[detached HEAD 8de0853] an honest mistake
+ 1 file changed, 1 insertion(+)
+```
+
+Running `west update` we can indeed see that we're losing the commit:
+
+```bash
+cd ../../
+$ west update
+...
+=== updating dummy (shared/dummy):
+--- dummy: fetching, need revision main
+From github.com:lmapii/practical-zephyr-t2-dummy
+ * branch            main       -> FETCH_HEAD
+Warning: you are leaving 1 commit behind, not connected to
+any of your branches:
+
+  8de0853 an honest mistake
+
+If you want to keep it by creating a new branch, this may be a good time
+to do so with:
+
+ git branch <new-branch-name> 8de0853
+
+HEAD is now at d22a413 updated test.txt
+...
+```
+
+The output that you see above is just Git's output. You can't actually restore the commit `8de0853` since West discarded it, the changes are really lost.
```

Comment:
I don't think this is right, things should be recoverable in this state. The commit will not be garbage collected immediately so the output recommending `git branch <new-branch-name> 8de0853` should work.

For the other case of committing to the `manifest-rev` branch, things are trickier but not impossible to recover. In my testing, I see west move the branch to the destination commit, but the changes to the previous state are staged allowing you to recover from there.

Reply:
Yep, I didn't try this! Seems to work, I'll update the article. Thanks!